### DISCUSSION TITLE
i don't understand why sqlalchemy.exc.InvalidRequestError: Table Already defined error is related with foreign key settings

### DISCUSSION THREAD

**Discussion by chanyoung1998** [sqlalchemy/sqlalchemy#10755] [Usage Questions] [answered]
I'm encountering the **sqlalchemy.exc.InvalidRequestError: Table Already defined error**, and I suspect it might be related to the table creation order because that error is solved when the module import sequnce between tables with foreign key relationships is changed or that foreign key relation is removed. However, I'm unsure about why this error is occurring. Could you provide insights into why this error might be happening and how to address it, especially in the context of table creation order and foreign key relationships in SQLAlchemy?

**Comment by zzzeek** [comment]
you would have a `Table` statement or equivalent declarative mapping that is being called twice, or multiple statements of the same `Table` (or equivalent declarative mapping) in the import sequence.  if you are using extensions like `automap`, that complicates things as well.

can't really say anything further without specifics

**Reply by chanyoung1998** [reply]
Until now, everything has been working well. The database used is PostgreSQL, and a Flask server is being utilized. However, recently, when I added a new table, I also added the model code related to that table. The added code is as follows, added to `model_1.py`. At the end of the code, I used the `create_all` function as follows:

```python
class AddedClass(Base):
    __tablename__ = 'added_class_name'
    __table__ = Table(__tablename__, Metadata, autoload_with=Engine, schema='user_etest')

    def to_dict(self):
        return {c.key: getattr(self, c.key) for c in inspect(self).mapper.column_attrs}

Base.metadata.create_all(Engine)
```

The existing code is as follows. Similarly, at the end of the code, I used the `create_all` function:

In `model_2.py`:

```python
class OriginalClass(Base):
    __tablename__ = 'original_class_name'
    __table__ = Table(__tablename__, Metadata, Column('id', Integer, primary_key=True),
                      autoload_with=Engine, schema='subject')

    def to_dict(self):
        return {c.key: getattr(self, c.key) for c in inspect(self).mapper.column_attrs}

Base.metadata.create_all(Engine)
```

And in `app.py`, I imported them and executed the code:

```python
import model_1
import model_2

# ~

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=5050, threaded=True)
```

The basic code structure is as follows: `AddedClass` has a foreign key that refers to the primary key of `OriginalClass`.


The error occurs in the code `__table__ = Table(__tablename__, Metadata, Column('id', Integer, primary_key=True), autoload_with=Engine, schema='subject')` within the OriginalClass definition.

I have double-checked that there are no duplicated table names, and I'm not using `automap`. Strangely, when I use pgAdmin to delete the foreign key column in `AddedClass` or change the import order of `model_1` and `model_2 ` to `model_2` and `model_1`, the error is resolved. I don't understand why this code is causing a `Table Already Defined` error.

**Reply by CaselIT** [reply]
Just remove the create all.
You are reflecting the tables from the db, so they are already there and there is nothing to create

**Reply by chanyoung1998** [reply]
yes , i did delete create_all but it still happens. It seems to be problematic when creating a table with a foreign key first. In my opinion, the issue might be occurring within the Table constructor. When the Table constructor is executed, does it also create the table linked by the foreign key? If not, it seems challenging to explain this error. I hope that is the reason...  What do you think?

**Reply by CaselIT** [reply, marked as answer]
Oh ok, sorry. When reflecting a table with a foreign key sqlalchemy will also reflect the other side table. This is likely what's happening here.
You can probably do:
- Reflect all tables beforehand, using `Metadata.reflect()` (you may need to call this multiple times if you have many schema to reflect). This may not be optimal if the db has many table that you don't care about. See https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.MetaData.reflect
- Set `extend_existing=True` in the table. See https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.Table.params.extend_existing

**Reply by chanyoung1998** [reply]
Thank you for you guys' reply!! It really helps me to solve my question!

**Reply by chanyoung1998** [reply]
https://docs.sqlalchemy.org/en/20/core/reflection.html
